File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
                
        # Use DFS preorder approach:
        # Time cplexity: O(n). Space: O(n)
        
        def dfsHelper(node, maxVal):
            if not node:
                return 0
            res = 1 if node.val >= maxVal else 0
            maxVal = max(maxVal, node.val)
            
            res += dfsHelper(node.left, maxVal)
            res += dfsHelper(node.right, maxVal)
            return res
        
        return dfsHelper(root, root.val)

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfsHelper(node *TreeNode, maxVal int) int {
    res := 0
    if node == nil {
        return 0
    }
    if node.Val >= maxVal {
        res = 1
    } else {
        res = 0
    }
    maxVal = max(maxVal, node.Val)
    res += dfsHelper(node.Left, maxVal)
    res += dfsHelper(node.Right, maxVal)
    
    return res
}
func goodNodes(root *TreeNode) int {
    return dfsHelper(root, root.Val)
}

// Time Complexity: O(n). Space Complexity: O(1)